### PR TITLE
Allow to specify a custom EventExecutorChooserFactory. Related to [#1…

### PR DESCRIPTION
**Pull request by normanmaurer**
…230]

Motivation:

Sometimes it may be benefitially for an user to specify a custom algorithm when choose the next EventExecutor/EventLoop.

Modifications:

Allow to specify a custom EventExecutorChooseFactory that allows to customize algorithm.

Result:

More flexible api.

### CODE REVIEW

File: transport/src/main/java/io/netty/channel/nio/NioEventLoopGroup.java [hunk 1]
```diff
@@ -84,6 +85,12 @@ public NioEventLoopGroup(int nThreads, Executor executor, final SelectorProvider
         super(nThreads, executor, selectorProvider, selectStrategyFactory);
     }
 
+    public NioEventLoopGroup(int nThreads, Executor executor, EventExecutorChooserFactory chooserFactory,
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoopGroup.java]: Will also add these to the other EvnetLoop impls once @trustin reviewed.

**Comment by normanmaurer** [comment]
@trustin PTAL

**Comment by normanmaurer** [comment]
@buchgr also please check as this is related to some work you done in the past :)

File: common/src/main/java/io/netty/util/concurrent/EventExecutorChooserFactory.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Factory that creates new {@link EventExecutorChooser}s.
+ */
+public interface EventExecutorChooserFactory {
+
+    /**
+     * Returns a new {@link EventExecutorChooser}.
+     */
+    EventExecutorChooser newChooser(EventExecutor[] executors);
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/EventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    private static boolean isPowerOfTwo(int val) {
+        return (val & -val) == val;
+    }
+
+    private static final class PowerOfTowEventExecutorChooser implements EventExecutorChooser {
+        private final AtomicInteger idx = new AtomicInteger();
+        private final EventExecutor[] executors;
+
+        PowerOfTowEventExecutorChooser(EventExecutor[] executors) {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        public EventExecutor next() {
+            return executors[idx.getAndIncrement() & executors.length - 1];
+        }
+    }
+
+    private static final class GenericEventExecutorChooser implements EventExecutorChooser {
+        private final AtomicInteger idx = new AtomicInteger();
+        private final EventExecutor[] executors;
+
+        GenericEventExecutorChooser(EventExecutor[] executors) {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ */
+public final class DefaultEventExecutorChooserFactory implements EventExecutorChooserFactory {
+
+    public static final DefaultEventExecutorChooserFactory INSTANCE = new DefaultEventExecutorChooserFactory();
+
+    private DefaultEventExecutorChooserFactory() { }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public EventExecutorChooser newChooser(EventExecutor[] executors) {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

**Comment by buchgr** [comment]
Looks good, just a few thoughts:

- Maybe mark as unstable api, until a few people have actually implemented a useful `EventExecutorChooser`.
- I assume `EventExecutorMetrics` will be added next?
- I recall that we used to have a `next(Channel ch)` method, as someone had the need to choose an executor passed on the IP subnet. Not sure if that's a good idea / useful.

File: transport/src/main/java/io/netty/channel/MultithreadEventLoopGroup.java [hunk 6]
```diff
@@ -57,6 +58,15 @@ protected MultithreadEventLoopGroup(int nThreads, ThreadFactory threadFactory, O
         super(nThreads == 0 ? DEFAULT_EVENT_LOOP_THREADS : nThreads, threadFactory, args);
     }
 
+    /**
+     * @see {@link MultithreadEventExecutorGroup#MultithreadEventExecutorGroup(int, Executor,
+     * EventExecutorChooserFactory, Object...)}
+     */
+    protected MultithreadEventLoopGroup(int nThreads, Executor executor, EventExecutorChooserFactory chooserFactory,
+                                     Object... args) {
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/MultithreadEventLoopGroup.java]: indentation

**Comment by trustin** [comment]
LGTM. +1 for marking it unstable.

**Comment by normanmaurer** [comment]
@buchgr yes I will add some kind of interface which expose some metrics as next step ;)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (b461c9d54c40b79403d21aa6980ae4c4abcb3283)

**Pull request closed without merging**